controllers/dataprotection/backuprepo_controller.go

```diff
@@ -0,0 +1,830 @@
+		metav1.ConditionTrue, ReasonDerivedObjectsDeleted, "")
+	if err != nil {
+		return fmt.Errorf("failed to update condition: %w", err)
+	}
+
+	// maintain mappers
+	r.secretRefMapper.removeRef(repo)
+	r.providerRefMapper.removeRef(repo)
+
+	return nil
+}
+
+func (r *BackupRepoReconciler) deletePVCs(reqCtx intctrlutil.RequestCtx, repo *dpv1alpha1.BackupRepo) (clear bool, err error) {
+	pvcList := &corev1.PersistentVolumeClaimList{}
+	if err := r.Client.List(reqCtx.Ctx, pvcList,
+		client.MatchingLabels(map[string]string{
+			dataProtectionBackupRepoKey: repo.Name,
+		})); err != nil {
+		return false, fmt.Errorf("failed to list PVCs: %w", err)
+	}
+
+	for _, pvc := range pvcList.Items {
+		if !isOwned(repo, &pvc) {
+			continue
+		}
+		reqCtx.Log.Info("deleting PVC", "name", pvc.Name, "namespace", pvc.Namespace)
+		if err := intctrlutil.BackgroundDeleteObject(r.Client, reqCtx.Ctx, &pvc); err != nil {
+			return false, err
+		}
+	}
+	// make sure all derived PVCs are deleted
+	clear = true
+	for _, pvc := range pvcList.Items {
+		if !isOwned(repo, &pvc) {
+			continue
+		}
+		err = r.Client.Get(reqCtx.Ctx, client.ObjectKeyFromObject(&pvc), &corev1.PersistentVolumeClaim{})
+		if !apierrors.IsNotFound(err) {
+			clear = false
+			break
+		}
+	}
+	return clear, nil
+}
+
+func (r *BackupRepoReconciler) deleteStorageClasses(reqCtx intctrlutil.RequestCtx, repo *dpv1alpha1.BackupRepo) error {
+	scList := &storagev1.StorageClassList{}
+	if err := r.Client.List(reqCtx.Ctx, scList,
+		client.MatchingLabels(map[string]string{
+			dataProtectionBackupRepoKey: repo.Name,
+		})); err != nil {
+		return fmt.Errorf("failed to list StorageClasses: %w", err)
+	}
+
+	for _, sc := range scList.Items {
+		if !isOwned(repo, &sc) {
+			continue
+		}
+		reqCtx.Log.Info("deleting StorageClass", "storageclass", sc.Name)
+		if err := intctrlutil.BackgroundDeleteObject(r.Client, reqCtx.Ctx, &sc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *BackupRepoReconciler) deleteSecrets(reqCtx intctrlutil.RequestCtx, repo *dpv1alpha1.BackupRepo) error {
+	secretList := &corev1.SecretList{}
+	if err := r.Client.List(reqCtx.Ctx, secretList,
+		client.MatchingLabels(map[string]string{
+			dataProtectionBackupRepoKey: repo.Name,
+		})); err != nil {
+		return fmt.Errorf("failed to list Secret: %w", err)
+	}
+
+	for _, secret := range secretList.Items {
+		if !isOwned(repo, &secret) {
+			continue
+		}
+		reqCtx.Log.Info("deleting Secret", "secret", client.ObjectKeyFromObject(&secret))
+		if err := intctrlutil.BackgroundDeleteObject(r.Client, reqCtx.Ctx, &secret); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *BackupRepoReconciler) mapBackupToRepo(obj client.Object) []ctrl.Request {
+	repoName, ok := obj.GetLabels()[dataProtectionBackupRepoKey]
+	if !ok {
+		return nil
+	}
+	if val := obj.GetAnnotations()[dataProtectionPVCCreatedAnnotationKey]; val == trueVal {
+		return nil
+	}
```

Comment:
need to filter failed backup?

Reply:
you can check if the pvc exists in backup controller. if not exists, add a label such as "dataprotection.kubeblocks.ip/pvc-not-exists". and can filter with this label? also can filter by `listAssociatedBackups`

Reply:
> need to filter failed backup?

done

> you can check if the pvc exists in backup controller. if not exists, add a label such as "dataprotection.kubeblocks.ip/pvc-not-exists". and can filter with this label? also can filter by `listAssociatedBackups`

done